domain/resource/state/resource.go

```diff
@@ -489,60 +472,107 @@ WHERE application_uuid = $resourceIdentity.application_uuid
 				charmRes, err := res.toCharmResource()
 				if err != nil {
 					return errors.Capture(err)
 				}
-				result.RepositoryResources = append(result.RepositoryResources, charmRes)
+				// Add potential resource.
+				potential = append(potential, charmRes)
 				continue
 			}
 
-			resId := resourceIdentity{UUID: res.UUID, ApplicationUUID: res.ApplicationUUID}
-
-			// Fetch units related to the resource.
-			var units []unitResource
-			err = tx.Query(ctx, getUnitStmt, resId).GetAll(&units)
-			if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
-				return errors.Capture(err)
-			}
-
 			r, err := res.toResource()
 			if err != nil {
 				return errors.Capture(err)
 			}
-			// Add each resource.
-			result.Resources = append(result.Resources, r)
-
-			// Sort by unit to generate unit resources.
-			for _, unit := range units {
-				unitRes, ok := resByUnit[coreunit.UUID(unit.UnitUUID)]
-				if !ok {
-					return errors.Errorf("unexpected unit %q linked to resource %q.", unit.UnitUUID, unitRes.Name)
-				}
-				ur, err := res.toResource()
+			// Add available resource.
+			available = append(available, r)
+		}
+		return nil
+	})
+	return potential, available, errors.Capture(err)
+}
+
+func (st *State) listUnitResources(
+	ctx context.Context,
+	applicationID application.ID,
+) ([]coreresource.UnitResources, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	// Prepare the application ID to query resources by application.
+	appID := resourceIdentity{
+		ApplicationUUID: applicationID.String(),
+	}
+
+	// Prepare the statement to get units related to a resource.
+	getApplicationUnitsStmt, err := st.Prepare(`
+		SELECT &unitUUIDAndName.*
+		FROM unit
+		WHERE application_uuid = $resourceIdentity.application_uuid
+			`, appID, unitUUIDAndName{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	// Prepare the statement to get resources linked to a unit
+	getResourcesStmt, err := st.Prepare(`
+		SELECT &resourceView.*
+		FROM v_unit_resource
+		WHERE unit_uuid = $unitUUIDAndName.uuid`, resourceView{}, unitUUIDAndName{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var unitResources []coreresource.UnitResources
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) (err error) {
+		// Units linked to the application.
+		var units []unitUUIDAndName
+
+		// Query to get all units for the given application.
+		err = tx.Query(ctx, getApplicationUnitsStmt, appID).GetAll(&units)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil // nothing found
+		}
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		// Process each unit linked to the application and get its resources.
+		for _, unit := range units {
+			var resourceViews []resourceView
+
+			err = tx.Query(ctx, getResourcesStmt, unit).GetAll(&resourceViews)
+			if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Errorf("get resources for unit %q: %w", unit.Name, err)
+			}
```

Comment:
question: Where is the expectation documented that ListResources contains a unit which isn't using the resources yet?